pcp/PCPDynamicColumn.c

```diff
@@ -0,0 +1,330 @@
+            metric->color = DYNAMIC_BLUE;
+         else if (String_eq(value, "cyan"))
+            metric->color = DYNAMIC_CYAN;
+         else if (String_eq(value, "magenta"))
+            metric->color = DYNAMIC_MAGENTA;
+         else if (String_eq(value, "yellow"))
+            metric->color = DYNAMIC_YELLOW;
+         else if (String_eq(value, "white"))
+            metric->color = DYNAMIC_WHITE;
+      } else if (String_eq(p, "label")) {
+         char* label = String_cat(value, ": ");
+         free_and_xStrdup(&metric->label, label);
+         free(label);
+      } else if (String_eq(p, "suffix")) {
+         free_and_xStrdup(&metric->suffix, value);
+      }
+   }
+}
+
+// Ensure a valid name for use in a PCP metric name and in htoprc
+static void PCPDynamicColumn_validateColumnName(char* key, const char* path, unsigned int line) {
+   char* p = key;
+   char* end = strrchr(key, ']');
+
+   if (end) {
+      *end = '\0';
+   } else {
+      char* note;
+      xAsprintf(&note,
+                "%s: no closing brace on column....... name at %s line %u\n\"%s\"",
+                pmGetProgname(), path, line, key);
+      errno = EINVAL;
+      CRT_fatalError(note);
+      free(note);
+   }
+
+   while (*p) {
+      if (p == key) {
+         if (!isalpha(*p) && *p != '_')
+            break;
+      } else {
+         if (!isalnum(*p) && *p != '_')
+            break;
+      }
+      p++;
+   }
+   if (*p != '\0') { /* badness */
+      char* note;
+      xAsprintf(&note,
+                "%s: invalid column....... name at %s line %u\n\"%s\"",
+                pmGetProgname(), path, line, key);
+      errno = EINVAL;
+      CRT_fatalError(note);
+      free(note);
+   } else { /* overwrite closing brace */
+      *p = '\0';
+   }
+}
+
+static PCPDynamicColumn* PCPDynamicColumn_new(PCPDynamicColumns* columns, const char* name) {
+   PCPDynamicColumn* column = xCalloc(1, sizeof(*column));
+   String_safeStrncpy(column->super.name, name, sizeof(column->super.name));
+   Hashtable_put(columns->table, ++columns->count, column);
+   return column;
+}
+
+static void PCPDynamicColumn_parseFile(PCPDynamicColumns* columns, const char* path) {
+   FILE* file = fopen(path, "r");
+   if (!file)
+      return;
+
+   PCPDynamicColumn* column = NULL;
+   unsigned int lineno = 0;
+   for (;;) {
+      char* line = String_readLine(file);
+      if (!line)
+         break;
+      lineno++;
+
+      /* cleanup whitespace, skip comment lines */
+      char* trimmed = String_trim(line);
+      free(line);
+      if (trimmed[0] == '#' || trimmed[0] == '\0') {
+         free(trimmed);
+         continue;
+      }
+
+      size_t n;
+      char** config = String_split(trimmed, '=', &n);
+      free(trimmed);
+      if (config == NULL)
+         continue;
+
+      char* key = String_trim(config[0]);
+      char* value = n > 1 ? String_trim(config[1]) : NULL;
+      if (key[0] == '[') {  /* new section heading - i.e. new column */
```

Comment:
Shouldn't the heading check for `[` appear before trying to split the line at `=`?

Reply:
Okay, that's a chance for cleanup of both locations in another PR then …